issue_1/contracts/WordRegistry.sol

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.4;
+
+import "hardhat/console.sol";
+
+import "./yield-utils-v2/contracts/access/Ownable.sol";
+
+/**
+@title Dictatorship-friendly word registry
+@author Owleksiy
+@notice This contract allows users to register/unregister (claim/unclaim) words
+Once a word is regiistered by a user, other users can't register it, unless the
+original user unregisters it.
+
+Additionally, the owner of the contract can expropriate any registered word
+ */
+contract WordRegistry is Ownable {
+    event OwnershipChange(string word, address previousOwner, address newOwner);
+    // word -> owner mapping
+    mapping(string => address) private word_to_owner;
+
+    constructor() {
+        console.log("Creating a registry, owner:", msg.sender);
+    }
+
+    function _is_owned_by_somebody(string memory word) private view returns (bool) {
+        return word_to_owner[word] != address(0);
+    }
+
+    function _transfer(
+        string memory word,
+        address old_owner,
+        address new_owner
+    ) private {
+        word_to_owner[word] = new_owner;
```

Comment:
You should check that either `old_owner` is the current owner of the name, or that the name is not registered.

Reply:
This function does not do any checks intentionally - it's internal and relies on callers to do the prep work. Will clarify with comments

Reply:
(for example, it's used in 'expropriate' which intentionally allows the owner to take over other user's words)

Reply:
When I split functions into internal and external parts, the external parts verify the user inputs and the initial state, but the internal ones verify data integrity.

That the function will let you transfer a name from Alice to Bob, and record the fact by emitting an event that says `OwnershipChange(name, Alice, Bob)`, when Alice never owned the name, seems to me to be asking for trouble.

Since you are probably familiar with ERC20, why not reusing the same patterns? One function to register a name, one function to transfer it, one function to unregister it.